packages/app/retainer/retainer.go

```diff
@@ -147,40 +174,77 @@ func (r *Retainer) setupEvents() {
 			cm.ConflictIDs = evt.ConflictIDs
 			cm.Unlock()
 		}
-	}, event.WithWorkerPool(r.workerPool))
+	}, event.WithWorkerPool(r.blockWorkerPool))
 
 	r.protocol.Events.Engine.Tangle.Booker.VirtualVoting.BlockTracked.Hook(func(block *virtualvoting.Block) {
 		if cm := r.createOrGetCachedMetadata(block.ID()); cm != nil {
 			cm.setVirtualVotingBlock(block)
 		}
-	}, event.WithWorkerPool(r.workerPool))
+	}, event.WithWorkerPool(r.blockWorkerPool))
 
 	congestionControl := func(block *scheduler.Block) {
 		if cm := r.createOrGetCachedMetadata(block.ID()); cm != nil {
 			cm.setSchedulerBlock(block)
 		}
 	}
-	r.protocol.Events.CongestionControl.Scheduler.BlockScheduled.Hook(congestionControl, event.WithWorkerPool(r.workerPool))
-	r.protocol.Events.CongestionControl.Scheduler.BlockDropped.Hook(congestionControl, event.WithWorkerPool(r.workerPool))
-	r.protocol.Events.CongestionControl.Scheduler.BlockSkipped.Hook(congestionControl, event.WithWorkerPool(r.workerPool))
+	r.protocol.Events.CongestionControl.Scheduler.BlockScheduled.Hook(congestionControl, event.WithWorkerPool(r.blockWorkerPool))
+	r.protocol.Events.CongestionControl.Scheduler.BlockDropped.Hook(congestionControl, event.WithWorkerPool(r.blockWorkerPool))
+	r.protocol.Events.CongestionControl.Scheduler.BlockSkipped.Hook(congestionControl, event.WithWorkerPool(r.blockWorkerPool))
 
 	r.protocol.Events.Engine.Consensus.BlockGadget.BlockAccepted.Hook(func(block *blockgadget.Block) {
 		cm := r.createOrGetCachedMetadata(block.ID())
 		cm.setAcceptanceBlock(block)
-	}, event.WithWorkerPool(r.workerPool))
+	}, event.WithWorkerPool(r.blockWorkerPool))
 
 	r.protocol.Events.Engine.Consensus.BlockGadget.BlockConfirmed.Hook(func(block *blockgadget.Block) {
 		if cm := r.createOrGetCachedMetadata(block.ID()); cm != nil {
 			cm.setConfirmationBlock(block)
 		}
-	}, event.WithWorkerPool(r.workerPool))
+	}, event.WithWorkerPool(r.blockWorkerPool))
 
 	r.protocol.Events.Engine.EvictionState.EpochEvicted.Hook(r.storeAndEvictEpoch)
+
+	r.protocol.Engine().NotarizationManager.Events.EpochCommitted.Hook(func(e *notarization.EpochCommittedDetails) {
+		if e.Commitment.Index() < r.protocol.Engine().EvictionState.LastEvictedEpoch() {
+			return
+		}
+
+		cd, _ := r.getCommitmentDetails(e.Commitment.Index())
```

Comment:
Why get it from storage? We should only commit each epoch once so you should always be able to create a new `CommitmentDetails` and then just `setCommitment` (don't forget to set the ID in here) and finally store it.